### ISSUE TITLE
Replace int-based X, Y, Z params with Vector3 / cCuboid / cBoundingBox

### ISSUE THREAD

**Issue by madmaxoft** [cuberite/cuberite#3853]
Currently the code sometimes uses three `int` coords for the X, Y, Z params, sometimes it uses a `Vector3`. Even worse, when specifying a cuboid, the code sometimes uses the `(MinX, MaxX, MinY, MaxY, MinZ, MaxZ)` ordering and sometimes the `(MinX, MinY, MinZ, MaxX, MaxY, MaxZ)` ordering.

We should replace most - if not all - of these with `Vector3` and `cCuboid` / `cBoundingBox`. If there are overloads for both, remove the `int`-based versions (but keep them in Lua API for compatibility reasons, possibly with a warning).

This should be done incrementally in small steps, not in one big PR, otherwise it risks never being done due to too many merge conflicts.

**Comment by madmaxoft** [comment]
Note: There's a poll in the forum for this: https://forum.cuberite.org/thread-2987.html

**Comment by peterbell10** [comment]
Posting this here because I want it to be visible after #3874 is merged but it's very much in response to that discussion.
I've written a little benchmark to help decide whether a `Vector3i` should be passed by value or by reference.  I can't say I'm 100% confident of these results because benchmarking something as small as a function call is bound to be hard but this is my best attempt anyway.
You can fiddle with the benchmark [here](http://quick-bench.com/wHRLcjKNcS8aOMKSZLOtllU2LRw).  I've tried to benchmark both methods in two ways: the breadth and depth benchmarks.  The idea being that the depth variant uses tail recursion and the vectors are not forced onto the stack unlike the breadth variants.

Here's one of the results from O2 (NB: breadth and depth aren't comparable):
![o2](https://user-images.githubusercontent.com/13238737/28757253-72f94f6a-7576-11e7-9ba1-fbb2c24272c2.png)

If my method is sound then when the values are on the stack, passing by const ref is very marginally faster but when they can be kept in registers, passing by value is more noticeably faster.

**Comment by worktycho** [comment]
Also, if the functions are inlined as they will be in a loop, it quite frankly doesn't make the blindest bit of difference. With the following use functions there is 5-7 prolog instructions as the actual addition is inlined. If we aren't in a loop then code loading with give more than enough latency to hide the loads.
```
ValVector3 SelfAdd(ValVector3 a_Vec, int Depth)
{
    for (int i = 0; i < 256*256; i++) {
        a_Vec = a_Vec + a_Vec;
    }
  return a_Vec;
}

RefVector3 SelfAdd(const RefVector3 & a_Vec, int Depth)
{
    RefVector3 vec = a_Vec;
    for (int i = 0; i < 256*256; i++) {
        vec = vec + vec;
    }
  return vec;
}```

**Comment by peterbell10** [comment]
In the context of cuberite there are quite a few virtual calls and functions defined in other translation units etc. that will prevent inlining so it's not completely irrelevant.  Although that being said, I'm sure you'd be extremely hard pressed to find a noticeable difference in anything other than synthetic micro-benchmarks.

**Comment by worktycho** [comment]
@peterbell10 But we avoid functions in other translation units in tight loops, and in non-loop code, the cost of loading the code will probably hide it. If you know any out of translation unit calls being made in tight loops other than the redstone simulator (I know about that one) I'd love to know. But we definitely have bigger issues to worry about.

**Comment by peterbell10** [comment]
When profiling the lighting code I noticed that `cPieceGeneratorBFSTree` spends a lot of time in `cCuboid::DoesIntersect` and that's just the first that comes to mind.  I do agree it's certainly not a pressing issue though.

**Comment by worktycho** [comment]
Hmm, maybe we should try optimizing DoesIntesect. Using `return (a_Min1 < a_Min2 & a_Max1 > a_Min2) | (a_Min1 >= a_Min2 & a_Min1 < a_Max2);` as the interval intersection, moving the DoesIntersect functions to bitwise and allowing the function to be inlined allows clang to produce much nicer code, potentially even vectorising it, but thats probably a different issue.

**Comment by lkolbly** [comment]
If it's guaranteed that a_Min1 <= a_Max1 and a_Min2 <= a_Max2, then DoIntervalsIntersect can be made even shorter:
```
bool DoIntervalsIntersectShort(int a_Min1, int a_Max1, int a_Min2, int a_Max2)
{
  return !((a_Min1 > a_Max2) || (a_Max1 < a_Min2));
}
```

**Comment by lkolbly** [comment]
Just made a PR to shorten it, #3883 (ooh, a palindrome). I tried building the generator performance tool, but got some compilation issues, so I'll have to look at it later.

**Comment by pwnorbitals** [comment]
Is there a list of places where XYZ replacement still needs to be done ?

**Comment by tonitch** [comment]
about vector3, I made a quick search and found really few of them missing (I just searched for "int x" in all sources files and checked em one by one)

here the list I found (might be wrong for some tho)
```
src/Generating/FinishGen.h#L420:	bool TryPlaceSpring(cChunkDesc & a_ChunkDesc, int x, int y, int z);
src/Generating/FinishGen.h#L449:	bool TrySpawnAnimals(cChunkDesc & a_ChunkDesc, int x, int y, int z, eMonsterType AnimalToSpawn);
src/Generating/FinishGen.cpp#L1519:bool cFinishGenFluidSprings::TryPlaceSpring(cChunkDesc & a_ChunkDesc, int x, int y, int z)

src/Entities/Entity.cpp#L1043:			int x, y, z;		# I'm wondering if we can put a vector3i here instead of creating a new struct.
src/Generating/FinishGen.cpp#L1530:		int x, y, z;		# Same here

src/Generating/MineShafts.cpp#L1030:		int x, y, z;		# Same for this one but there is a eDirection so might not be the case

src/ChunkDef.h#L210:	inline static size_t MakeIndex(int x, int y, int z)		# Already done but the deprecated fonction is still in the code
src/ChunkDef.h#L318:	static NIBBLETYPE GetNibble(const NIBBLETYPE * a_Buffer, int x, int y, int z)	# Not this one
src/ChunkDef.h#L467:	int x;						# Here, I don't know if it's usefull but maybe we can store a vector3 instead of x, y, z? don't know if it suits the object tho
```
